### ISSUE TITLE
Behavioural Invariants for Whiley

### ISSUE THREAD

**Issue by DavePearce** [Whiley/WhileyCompiler#757]
_This is an initial proposal which I might work up into an RFC_

The proposal is to add the concept of a `varying` clause, which describes the permitted states of a variable before and after an operation.  The following illustrates:

```
type upint is (int x) 
// Can never be negative
where x >= 0 
// Each change must increase value
varying x > x'
```

With this declaration, the following is permitted:

```
upint x = 0
x = x + 1
```

However, this is not permitted:

```
upint x = 0
x = 0
```

And neither is this:

```
upint x = 0
x = x - 1
```

The distinction between the `where` clause and the `varying` clause is subtle.  The former must hold true for every value held.  The latter must only hold true every time the value is changed.  Thus, for example, initialisers are exempt from the `varying` clause.  This means the following has meaning:

```
type gupint is (int x) varying x >= 0
```

Under this declaration, the following is permitted:

```
gupint x = - 1
x = 0
```

Basically, an instance of `gupint` can be initialised with anything but at the first assignment must be assigned something non-negative.

## Observations

Some initial observations:

- `varying` clauses are most useful for variables of reference type
- `varying` clauses might be useful for handling termination (e.g. of loops)

**Comment by DavePearce** [comment]
*NOTE:* this may be related to the concept of _history_ constraints.  This might be an example here:

ftp://nozdr.ru/biblio/kolxo3/Cs/CsLn/Programming%20Languages%20and%20Systems,%202%20conf.,%20APLAS%202004(LNCS3302,%20Springer,%202004)(ISBN%203540237240)(463s).pdf#page=117

**Comment by DavePearce** [comment]
Migrated to https://github.com/Whiley/RFCs/issues/37
